zebra-consensus/src/primitives/halo2.rs

```
@@ -0,0 +1,281 @@
+    ///
+    /// This is useful (in combination with `Item::clone`) for implementing
+    /// fallback logic when batch verification fails.
+    pub fn verify_single(&self, vk: &VerifyingKey) -> Result<(), halo2::plonk::Error> {
+        self.proof.verify(vk, &self.instances[..])
+    }
+}
+
+#[derive(Default)]
+pub struct BatchVerifier {
+    queue: Vec<Item>,
+}
+
+impl BatchVerifier {
+    pub fn queue(&mut self, item: Item) {
+        self.queue.push(item);
+    }
+
+    pub fn verify<R: RngCore + CryptoRng>(
+        self,
+        _rng: R,
+        vk: &VerifyingKey,
+    ) -> Result<(), halo2::plonk::Error> {
+        for item in self.queue {
+            item.verify_single(vk)?;
+        }
+
+        Ok(())
+    }
+}
+
+// === END TEMPORARY BATCH HALO2 SUBSTITUTE ===
+
+impl From<zebra_chain::orchard::ShieldedData> for Item {
+    fn from(shielded_data: zebra_chain::orchard::ShieldedData) -> Item {
+        use orchard::{circuit, note, primitives::redpallas, tree, value};
+
+        let anchor = tree::Anchor::from_bytes(shielded_data.shared_anchor.into()).unwrap();
+
+        let enable_spend = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_SPENDS);
+        let enable_output = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_OUTPUTS);
+
+        let instances = shielded_data
+            .actions()
+            .map(|action| {
+                circuit::Instance::from_parts(
+                    anchor,
+                    value::ValueCommitment::from_bytes(&action.cv.into()).unwrap(),
+                    note::Nullifier::from_bytes(&action.nullifier.into()).unwrap(),
+                    redpallas::VerificationKey::<redpallas::SpendAuth>::try_from(<[u8; 32]>::from(
+                        action.rk,
+                    ))
+                    .unwrap(),
+                    note::ExtractedNoteCommitment::from_bytes(&action.cm_x.into()).unwrap(),
```

I suggest using `expect` instead of `unwrap` with the reason it's OK to ignore errors (because we already validated them when parsing the tx, right?)

---

Updated some here: 031836edf96d6967234dcb4086474ff3cca32a01